File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/MaskCharacter.java
package com.vicmikhailau.maskededittext;


abstract class MaskCharacter {
    abstract public boolean isValidCharacter(char ch);

    public char processCharacter(char ch) {
        return ch;
    }

    public boolean isPrepopulate() {
        return false;
    }

}

class DigitCharacter extends MaskCharacter {
    @Override
    public boolean isValidCharacter(char ch) {
        return Character.isDigit(ch);
    }

}

class UpperCaseCharacter extends MaskCharacter {
    @Override
    public boolean isValidCharacter(char ch) {
        return Character.isUpperCase(ch);
    }

    @Override
    public char processCharacter(char ch) {
        return Character.toUpperCase(ch);
    }
}


class LowerCaseCharacter extends MaskCharacter {
    @Override
    public boolean isValidCharacter(char ch) {
        return Character.isLowerCase(ch);
    }

    @Override
    public char processCharacter(char ch) {
        return Character.toLowerCase(ch);
    }
}

class AlphaNumericCharacter extends MaskCharacter {
    @Override
    public boolean isValidCharacter(char ch) {
        return Character.isLetterOrDigit(ch);
    }
}

class LetterCharacter extends MaskCharacter {
    @Override
    public boolean isValidCharacter(char ch) {
        return Character.isLetter(ch);
    }
}

class HexCharacter extends MaskCharacter {
    private static final String HEX_CHARS = "0123456789ABCDEF";

    @Override
    public boolean isValidCharacter(char ch) {
        return Character.isLetterOrDigit(ch) && HEX_CHARS.indexOf(Character.toUpperCase(ch)) != -1;
    }

    @Override
    public char processCharacter(char ch) {
        return Character.toUpperCase(ch);
    }
}

class LiteralCharacter extends MaskCharacter {
    private Character character;

    LiteralCharacter() {
        character = null;
    }

    LiteralCharacter(char ch) {
        character = ch;
    }

    @Override
    public boolean isValidCharacter(char ch) {
        return character == null || character == ch;
    }


    @Override
    public char processCharacter(char ch) {
        if (character != null)
            return character;
        return ch;
    }

    public boolean isPrepopulate() {
        return character != null;
    }
}

File: MaskedEditText/src/main/java/com/vicmikhailau/maskededittext/MaskCharacter.kt
package com.vicmikhailau.maskededittext


abstract class MaskCharacter {

    open val isPrepopulate: Boolean
        get() = false

    abstract fun isValidCharacter(ch: Char): Boolean

    open fun processCharacter(ch: Char): Char {
        return ch
    }

}

internal class DigitCharacter : MaskCharacter() {
    override fun isValidCharacter(ch: Char): Boolean {
        return Character.isDigit(ch)
    }

}

internal class UpperCaseCharacter : MaskCharacter() {
    override fun isValidCharacter(ch: Char): Boolean {
        return Character.isUpperCase(ch)
    }

    override fun processCharacter(ch: Char): Char {
        return Character.toUpperCase(ch)
    }
}


internal class LowerCaseCharacter : MaskCharacter() {
    override fun isValidCharacter(ch: Char): Boolean {
        return Character.isLowerCase(ch)
    }

    override fun processCharacter(ch: Char): Char {
        return Character.toLowerCase(ch)
    }
}

internal class AlphaNumericCharacter : MaskCharacter() {
    override fun isValidCharacter(ch: Char): Boolean {
        return Character.isLetterOrDigit(ch)
    }
}

internal class LetterCharacter : MaskCharacter() {
    override fun isValidCharacter(ch: Char): Boolean {
        return Character.isLetter(ch)
    }
}

internal class HexCharacter : MaskCharacter() {

    override fun isValidCharacter(ch: Char): Boolean {
        return Character.isLetterOrDigit(ch) && HEX_CHARS.indexOf(Character.toUpperCase(ch)) != -1
    }

    override fun processCharacter(ch: Char): Char {
        return Character.toUpperCase(ch)
    }

    companion object {
        private val HEX_CHARS = "0123456789ABCDEF"
    }
}

internal class LiteralCharacter : MaskCharacter {
    private var character: Char? = null

    override val isPrepopulate: Boolean
        get() = character != null

    constructor() {
        character = null
    }

    constructor(ch: Char) {
        character = ch
    }

    override fun isValidCharacter(ch: Char): Boolean {
        return character == null || character == ch
    }


    override fun processCharacter(ch: Char): Char {
        return if (character != null) character!! else ch
    }
}

